# Comparing `tmp/primal_page-1.5.1.tar.gz` & `tmp/primal_page-1.6.0.tar.gz`

## filetype from file(1)

```diff
@@ -1 +1 @@
-gzip compressed data, was "primal_page-1.5.1.tar", max compression
+gzip compressed data, was "primal_page-1.6.0.tar", max compression
```

## Comparing `primal_page-1.5.1.tar` & `primal_page-1.6.0.tar`

### file list

```diff
@@ -1,10 +1,12 @@
--rw-r--r--   0        0        0    17347 2024-04-01 22:08:40.025579 primal_page-1.5.1/README.md
--rw-r--r--   0        0        0       22 2024-04-01 22:09:43.252047 primal_page-1.5.1/primal_page/__init__.py
--rw-r--r--   0        0        0     6457 2024-03-28 12:37:49.209826 primal_page-1.5.1/primal_page/bedfiles.py
--rw-r--r--   0        0        0     8570 2024-03-28 12:36:56.862086 primal_page-1.5.1/primal_page/build_index.py
--rw-r--r--   0        0        0     3106 2024-03-28 12:28:04.284143 primal_page-1.5.1/primal_page/download.py
--rw-r--r--   0        0        0    28125 2024-04-01 21:00:06.373740 primal_page-1.5.1/primal_page/main.py
--rw-r--r--   0        0        0     1959 2024-04-01 21:15:05.696388 primal_page-1.5.1/primal_page/modify.py
--rw-r--r--   0        0        0     5574 2024-03-29 09:47:36.111159 primal_page-1.5.1/primal_page/schemas.py
--rw-r--r--   0        0        0      996 2024-04-01 22:09:34.860450 primal_page-1.5.1/pyproject.toml
--rw-r--r--   0        0        0    17980 1970-01-01 00:00:00.000000 primal_page-1.5.1/PKG-INFO
+-rw-r--r--   0        0        0    17347 2024-04-01 22:08:40.025579 primal_page-1.6.0/README.md
+-rw-r--r--   0        0        0       22 2024-05-14 20:17:44.564268 primal_page-1.6.0/primal_page/__init__.py
+-rw-r--r--   0        0        0     9497 2024-05-04 12:17:24.470509 primal_page-1.6.0/primal_page/bedfiles.py
+-rw-r--r--   0        0        0     8570 2024-05-14 20:09:36.789671 primal_page-1.6.0/primal_page/build_index.py
+-rw-r--r--   0        0        0     3734 2024-05-04 12:16:48.895640 primal_page-1.6.0/primal_page/dev.py
+-rw-r--r--   0        0        0      805 2024-05-12 23:24:00.260665 primal_page-1.6.0/primal_page/doi.py
+-rw-r--r--   0        0        0     5071 2024-05-04 11:14:22.628600 primal_page-1.6.0/primal_page/download.py
+-rw-r--r--   0        0        0    15751 2024-05-14 18:29:13.197545 primal_page-1.6.0/primal_page/main.py
+-rw-r--r--   0        0        0    14315 2024-05-14 19:47:58.958929 primal_page-1.6.0/primal_page/modify.py
+-rw-r--r--   0        0        0     7077 2024-05-14 19:52:22.795325 primal_page-1.6.0/primal_page/schemas.py
+-rw-r--r--   0        0        0      996 2024-05-14 20:17:22.345341 primal_page-1.6.0/pyproject.toml
+-rw-r--r--   0        0        0    17980 1970-01-01 00:00:00.000000 primal_page-1.6.0/PKG-INFO
```

### Comparing `primal_page-1.5.1/README.md` & `primal_page-1.6.0/README.md`

 * *Files identical despite different names*

### Comparing `primal_page-1.5.1/primal_page/bedfiles.py` & `primal_page-1.6.0/primal_page/bedfiles.py`

 * *Files 22% similar despite different names*

```diff
@@ -166,25 +166,44 @@
         pool: int,
         strand: str,
         seq: str,
     ):
         self.chrom = chrom
         self.start = start
         self.end = end
-        self.name = primername
+        self.primername = primername
         self.pool = pool
         self.strand = strand
         self.seq = seq
 
+        # Validate the primername
+        if determine_primername_version(primername) == PrimerNameVersion.INVALID:
+            raise ValueError(f"Invalid primername: {primername}")
+
         # Autogenerated fields
-        self.primername_version = determine_primername_version(primername)
+        self.prefix = primername.split("_")[0]
         self.amplicon_number = int(primername.split("_")[1])
+        self.pn_direction = primername.split("_")[2]
+        self.primernumber: int | None = None
 
     def __str__(self):
-        return f"{self.chrom}\t{self.start}\t{self.end}\t{self.name}\t{self.pool}\t{self.strand}\t{self.seq}"
+        return f"{self.chrom}\t{self.start}\t{self.end}\t{self.primername}\t{self.pool}\t{self.strand}\t{self.seq}"
+
+    def parsed_primername(self, primernumber: int) -> str:
+        return "_".join(
+            [
+                self.prefix,
+                str(self.amplicon_number),
+                self.pn_direction,
+                str(primernumber),
+            ]
+        )
+
+    def parsed_bedline(self) -> str:
+        return f"{self.chrom}\t{self.start}\t{self.end}\t{self.primername if self.primernumber is None else self.parsed_primername(self.primernumber)}\t{self.pool}\t{self.strand}\t{self.seq}"
 
 
 def read_bedlines(bedfilepath: pathlib.Path) -> tuple[list[BedLine], list[str]]:
     """
     Reads a bed file and returns a list of BedLine objects.
 
     :return: bedfile_list
@@ -222,7 +241,70 @@
             # Header line
             if line.startswith("#"):
                 bedfile_header.append(line)
             elif line:  # If not empty
                 bedfile_list.append(line.split("\t"))
 
     return bedfile_list, bedfile_header
+
+
+def regenerate_v3_bedfile(bedfile: pathlib.Path) -> str:
+    """
+    THIS FUNCTION MODIFIES THE BEDFILE. ONLY USE WHEN UPDATING THE BEDFILE VERSION.
+
+    Convert a v2 bedfile to a v3 bedfile.
+    :param bedfile: The v2 bedfile
+    :return: The v3 bedfile
+    """
+
+    # Read in the bedfile
+    bedlines, bedheader = read_bedlines(bedfile)
+
+    # GUARD: If the bedfile is already v3, return the bedfile
+    if determine_bedfile_version(bedfile) == BedfileVersion.V3:
+        return bedfile.read_text()
+
+    # Sort the bedlines by chrom and amplicon number
+    bedlines = sorted(bedlines, key=lambda x: (x.chrom, x.amplicon_number))
+
+    # Resolve alt primers
+    # generate primerpairs
+    # {chromname: {ampliconnumber: ([fbedlines], [rbedlines])}}
+    sorted_bedlines: dict[str, dict[int, tuple[list[BedLine], list[BedLine]]]] = {}
+    for line in bedlines:
+        if line.chrom not in sorted_bedlines:
+            sorted_bedlines[line.chrom] = {}
+
+        if line.amplicon_number not in sorted_bedlines[line.chrom]:
+            sorted_bedlines[line.chrom][line.amplicon_number] = ([], [])
+
+        match line.strand:
+            case "+":
+                sorted_bedlines[line.chrom][line.amplicon_number][0].append(line)
+            case "-":
+                sorted_bedlines[line.chrom][line.amplicon_number][1].append(line)
+            case _:
+                raise ValueError(f"Invalid strand: {line.__str__()}")
+
+    bedfile_str_list = bedheader
+
+    # Calculate the primernumbers for alt primers
+    for amplicon_number_dict in sorted_bedlines.values():
+        for direction_list in amplicon_number_dict.values():
+            fprimers, rprimers = direction_list
+
+            # Sort and renumber the primers
+            fprimers = sorted(fprimers, key=lambda x: x.seq)
+            for i, line in enumerate(fprimers):
+                line.primernumber = i + 1
+
+                # Add the line to the bedfile
+                bedfile_str_list.append(line.parsed_bedline())
+
+            rprimers = sorted(rprimers, key=lambda x: x.seq)
+            for i, line in enumerate(rprimers):
+                line.primernumber = i + 1
+
+                # Add the line to the bedfile
+                bedfile_str_list.append(line.parsed_bedline())
+
+    return "\n".join(bedfile_str_list)
```

### Comparing `primal_page-1.5.1/primal_page/build_index.py` & `primal_page-1.6.0/primal_page/build_index.py`

 * *Files 0% similar despite different names*

```diff
@@ -228,15 +228,15 @@
     # Update the github commit
     if git_commit is not None:
         json_dict["github-commit-sha"] = (
             git_commit  # This is added so that the index.json changes when the github commit changes
         )
 
     with open(parent_dir / "index.json", "w") as f:
-        json.dump(json_dict, f, indent=4, sort_keys=True)
+        json.dump(json_dict, f, indent=1, sort_keys=True)
 
     return True
 
 
 if __name__ == "__main__":
     server_url = sys.argv[1]
     repo_url = sys.argv[2]
```

### Comparing `primal_page-1.5.1/primal_page/schemas.py` & `primal_page-1.6.0/primal_page/schemas.py`

 * *Files 26% similar despite different names*

```diff
@@ -3,18 +3,19 @@
 from typing import Annotated
 
 from pydantic import BaseModel, PositiveInt
 from pydantic.functional_validators import AfterValidator
 
 from primal_page.bedfiles import BedfileVersion
 
-INFO_SCHEMA = "v1.3.0"
+INFO_SCHEMA = "v2.0.0"
 
 SCHEMENAME_PATTERN = r"^[a-z0-9][a-z0-9-]*[a-z0-9]$"
 VERSION_PATTERN = r"^v\d+\.\d+\.\d+$"
+DOI_PATTERN = r"^(https:\/\/|dx.)*(doi.org\/|doi:)10.\d+\/.*$"
 
 
 class PrimerClass(Enum):
     PRIMERSCHEMES = "primerschemes"
 
 
 class SchemeStatus(Enum):
@@ -38,15 +39,15 @@
     if not re.match(SCHEMENAME_PATTERN, schemename):
         raise ValueError(
             f"Invalid schemename: {schemename}. Must only contain a-z, 0-9, and -. Cannot start or end with -"
         )
     return schemename
 
 
-def not_empty(x: list | set) -> list | set:
+def not_empty(x: list | set | str) -> list | set | str:
     if len(x) == 0:
         raise ValueError("Cannot be empty")
     return x
 
 
 class Collection(Enum):
     # Authors
@@ -59,14 +60,56 @@
     CLINAL_ISOLATES = "CLINAL-ISOLATES"
     # Coverage types
     WHOLE_GENOME = "WHOLE-GENOME"
     PANEL = "PANEL"
     MULTI_TARGET = "MULTI-TARGET"
 
 
+class Links(BaseModel):
+    protocals: list[str] = []
+    validation: list[str] = []
+    homepage: list[str] = []
+    vendors: list[str] = []
+    misc: list[str] = []
+
+    def getattr(self, link_type: str):
+        """
+        Get the list of links
+        Raises:
+        -   AttributeError if the link_type is not in the model_fields
+        """
+        return getattr(self, link_type)
+
+    def append_link(
+        self,
+        link_type: str,
+        link: str,
+    ):
+        """
+        Add a link to the correct list
+        Raises:
+        -   AttributeError if the link_type is not in the model_fields
+        """
+        # Append the link to the correct list
+        getattr(self, link_type).append(link)
+
+    def remove_link(
+        self,
+        link_type: str,
+        link: str,
+    ):
+        """
+        Remove a the first occourance of the link from the correct list
+        Raises:
+        -   ValueError if the link is not in the list
+        -   AttributeError if the link_type is not in the model_fields
+        """
+        getattr(self, link_type).remove(link)
+
+
 class Info(BaseModel):
     ampliconsize: PositiveInt
     schemeversion: Annotated[str, AfterValidator(validate_schemeversion)]
     schemename: Annotated[str, AfterValidator(validate_schemename)]
     primer_bed_md5: str
     reference_fasta_md5: str
     status: SchemeStatus
@@ -74,18 +117,20 @@
     authors: Annotated[list[str], AfterValidator(not_empty)]
     algorithmversion: str
     species: Annotated[set[int | str], AfterValidator(not_empty)]
     license: str = "CC BY-SA 4.0"
     primerclass: PrimerClass = PrimerClass.PRIMERSCHEMES
     infoschema: str = INFO_SCHEMA
     articbedversion: BedfileVersion
+    collections: set[Collection] = set()
+    links: Links = Links()
     # Add the optional fields
     description: str | None = None
     derivedfrom: str | None = None
-    collections: set[Collection] = set()
+    contactinfo: str | None = None
 
     def add_collection(self, collection: Collection):
         """Add a collection to the collections set"""
         self.collections.add(collection)
 
     def remove_collection(self, collection: Collection):
         """
@@ -161,23 +206,33 @@
         """Change the derivedfrom"""
         self.derivedfrom = new_derivedfrom
 
     def change_description(self, new_description: str):
         """Change the description"""
         self.description = new_description
 
+    def change_contactinfo(self, new_contactinfo: str | None):
+        """Change the contactinfo"""
+        self.contactinfo = new_contactinfo
+
 
 if __name__ == "__main__":
     info = Info(
         ampliconsize=400,
         schemeversion="v0.0.0",
         schemename="test",
         primer_bed_md5="hello",
         reference_fasta_md5="world",
         status=SchemeStatus.DRAFT,
         citations=set(),
         authors=["artic"],
         algorithmversion="test",
-        species=set("sars-cov-2"),
+        species=set(["sars-cov-2"]),
         articbedversion=BedfileVersion.V3,
         collections=set(),
+        links=Links(),
     )
+
+    info_json = info.model_dump_json()
+    print(info_json)
+
+    info2 = Info.model_validate_json(info_json)
```

### Comparing `primal_page-1.5.1/pyproject.toml` & `primal_page-1.6.0/pyproject.toml`

 * *Files 12% similar despite different names*

```diff
@@ -1,10 +1,10 @@
 [tool.poetry]
 name = "primal-page"
-version = "1.5.1"
+version = "1.6.0"
 description = "Tooling and schemas for the primerschemes index"
 authors = ["ChrisKent <chrisgkent@gmail.com>"]
 readme = "README.md"
 repository = "https://github.com/ChrisgKent/primal-page"
 
 [tool.poetry.dependencies]
 python = "^3.10"
```

### Comparing `primal_page-1.5.1/PKG-INFO` & `primal_page-1.6.0/PKG-INFO`

 * *Files 0% similar despite different names*

```diff
@@ -1,10 +1,10 @@
 Metadata-Version: 2.1
 Name: primal-page
-Version: 1.5.1
+Version: 1.6.0
 Summary: Tooling and schemas for the primerschemes index
 Home-page: https://github.com/ChrisgKent/primal-page
 Author: ChrisKent
 Author-email: chrisgkent@gmail.com
 Requires-Python: >=3.10,<4.0
 Classifier: Programming Language :: Python :: 3
 Classifier: Programming Language :: Python :: 3.10
```

